modules/training.py

```diff
@@ -245,11 +247,35 @@ def do_train(lora_name: str, always_override: bool, save_steps: int, micro_batch
     shared.tokenizer.pad_token_id = 0
     shared.tokenizer.padding_side = "left"
 
+    def encode(text):
+        return shared.tokenizer.encode(text, truncation=True, max_length=cutoff_len)
+
     def tokenize(prompt):
-        result = shared.tokenizer(prompt, truncation=True, max_length=cutoff_len + 1, padding="max_length")
+
+        if train_only_after == '' or train_only_after not in prompt:
+            input_ids = encode(prompt)
+            input_ids = [shared.tokenizer.pad_token_id] * (cutoff_len - len(input_ids)) + input_ids
+            labels = [1] * len(input_ids)
+
+        else:
+            ind = prompt.index(train_only_after) + len(train_only_after)
+            before_tokens = encode(prompt[:ind])[:-1]
+            after_tokens = encode(prompt[ind:])[1:]
+
+            full_length = len(after_tokens) + len(before_tokens)
+            if full_length > cutoff_len:
+                after_tokens = after_tokens[:cutoff_len - len(before_tokens)]
```

Comment:
Could there be an edge case where `len(before_tokens) > cutoff_len`?

Reply:
No, that's guaranteed against on line 251 `max_length=cutoff_len` in the tokenizer encode call.